_posts/2024-01-21-comparing-cas-and-karpenter.markdown

```diff
@@ -0,0 +1,274 @@
+instances which might fit the pod (or group of pods) and then choose the best
+option based on user preferences.
+
+Allowing the application to understand provisioning instead of scaling also
+lends itself to more dynamic discovery of instance types and categories at run
+time. This means that as cloud inventory changes, Karpenter can react to those
+changes automatically and make moment-to-moment decisions about market
+availability and pricing. By contrast, to emulate this behavior using the CAS
+with Cluster API would require the user to be updating the
+`InfrastructureMachineTemplate` resources in their cluster as well as updating
+the associated scalable types on a continual basis.
+
+## When to activate, pending or unscheduled?
+
+Another subtle point of difference that flows from the notion of scaling
+versus provisioning is the conditions under which these applications will jump
+into action.
+
+When there are pending pods in a cluster, this means that there
+is a node available which could receive the pod but that it currently does not
+have allocatable capacity. In these cases, when configured, the CAS and
+Karpenter both have the ability to  add more nodes of the type that will accept
+the workload.
+
+When there are unschedulable pods in a cluster, this means that there are no
+nodes that can satisfy the requirements of the pod. In these cases the CAS
+will only act if it has a node group which could possibly make a node to
+schedule that pod. This scenario can happen when the CAS is configured to
+have node groups of size zero and thus there are no nodes in the cluster which
+could schedule the pod, but the CAS knows how to make that type of node.
+
+Karpenter, by contrast, is configured by the user to instruct the provisioning
+of nodes based on pod constraints. In the case of an unschedulable pod,
+Karpenter will refer to its provisioners, and will then request several
+instance types which could satisfy the pod requirements. Karpenter can then
+choose the best (by cost, resource, availability, etc) instance type to create
+as the pod is being requested.
+
+This is not to say that the CAS won't also look at several instance types when
+deciding which to make, but we need to look at the details a little closer to
+understand that choice. In Karpenter, having the ability to calculate instance
+types based on the workload constraints means that it can request a broad
+range of instances (limited by user configuration) from the infrastructure. By
+contrast, CAS can be configured with many instance types and when presented
+with a decision about which type of instance to request for a specific workload
+it will use the user's preference (up to and including [custom code][cas-grpc])
+to make that choice. The CAS is limited by its node group configurations,
+which are different on each platform, and may or may not support dynamic
+instance discovery and full resource and cost data.
```

Comment:
> I think this is saying CAS expects the user to explicitly tell it about node types via a node group, but Karpenter checks the pod resource requests and then the cloud's inventory to find a match. Is that right? 

more or less, yes.

> Does CAS also look at pod resource requests to determine which node group to use to allocate a new node?

yes, CAS will use the kube scheduler and a set of fake nodes (the nodes it could make) to see if the pods would be resolved to any of the nodes that it can make. then it makes a decision based on the configured expander to determine the final node group.